File: src/main/scala/ru/tinkoff/gatling/amqp/request/RequestReplyDslBuilderMessage.scala
package ru.tinkoff.gatling.amqp.request

import io.gatling.core.config.GatlingConfiguration
import io.gatling.core.session.Expression
import ru.tinkoff.gatling.amqp.action.RequestReplyBuilder

import java.nio.charset.Charset

case class RequestReplyDslBuilderMessage(
    requestName: Expression[String],
    destination: AmqpExchange,
    replyDest: AmqpExchange,
    setReplyTo: Boolean,
    messageSelector: Option[String],
    configuration: GatlingConfiguration
) {

  /**
    * Add a reply queue
    */
  def replyExchange(name: Expression[String]): RequestReplyDslBuilderMessage = replyDestination(AmqpQueueExchange(name))
  private def replyDestination(destination: AmqpExchange)                    = this.copy(replyDest = destination)
  def noReplyTo: RequestReplyDslBuilderMessage                               = this.copy(setReplyTo = false)

  def textMessage(text: Expression[String], charset: Charset = configuration.core.charset): RequestReplyDslBuilder =
    message(TextAmqpMessage(text, charset))

  def bytesMessage(bytes: Expression[Array[Byte]]): RequestReplyDslBuilder = message(BytesAmqpMessage(bytes))

  private def message(mess: AmqpMessage) =
    RequestReplyDslBuilder(
      AmqpAttributes(requestName, destination, messageSelector, mess),
      RequestReplyBuilder(_, replyDest, setReplyTo, configuration)
    )
}

File: src/main/java/ru/tinkoff/gatling/amqp/javaapi/request/RequestReplyDslBuilderMessage.java
package ru.tinkoff.gatling.amqp.javaapi.request;

import java.nio.charset.Charset;

import static io.gatling.javaapi.core.internal.Expressions.*;

public class RequestReplyDslBuilderMessage {
    private ru.tinkoff.gatling.amqp.request.RequestReplyDslBuilderMessage wrapped;

    public RequestReplyDslBuilderMessage(ru.tinkoff.gatling.amqp.request.RequestReplyDslBuilderMessage wrapped){
        this.wrapped = wrapped;
    }

    public RequestReplyDslBuilderMessage replyExchange(String name){
        this.wrapped = wrapped.replyExchange(toStringExpression(name));
        return this;
    }

    public RequestReplyDslBuilderMessage noReplyTo(){
        this.wrapped = wrapped.noReplyTo();
        return this;
    }

    public RequestReplyDslBuilder textMessage(String text){
        return textMessage(text, io.gatling.core.Predef.configuration().core().charset());
    }

    public RequestReplyDslBuilder textMessage(String text, Charset charset){
        return new RequestReplyDslBuilder(wrapped.textMessage(toStringExpression(text), charset));
    }

    public RequestReplyDslBuilder bytesMessage(byte[] bytes){
        return new RequestReplyDslBuilder(wrapped.bytesMessage(toStaticValueExpression(bytes)));
    }

    public RequestReplyDslBuilder bytesMessage(String el){
        return new RequestReplyDslBuilder(wrapped.bytesMessage(toBytesExpression(el)));
    }
}
